### PR TITLE
chore: add totp pairing flow documentation

### PR DESCRIPTION
**Pull request by vinistevam**
# Overview

Add TOTP pairing flow documentation

See [here](https://github.com/MetaMask/metamask-desktop/blob/chore/568-add-totp-pairing-flow-documentation/docs/totp-pairing-flow.md) for markdown.

### CODE REVIEW

**Comment by vinistevam** [comment]
I have read the CLA Document and I hereby sign the CLA

File: docs/totp-pairing-flow.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        MME->>MMD: start a limited connection
+        activate MMD
+        Note over MME,MMD: - Create WebSocket<br/>- Encrypt Stream<br/>- Perform Handshake
+        MMD->>MME: return connection established
+        deactivate MMD
+        Note over MME: Generates TOTP
+    MME->>User: display Pairing Page to the user
+    User->>MMD: submit TOTP (6 digits) displayed in the MME
+    MMD->>MME: send TOTP to MME
+    Note over MME: Validate TOTP<br>Create Pairing Key
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:23]:
```suggestion
    Note over MME: - Validate TOTP<br/>- Create Pairing Key
```

File: docs/totp-pairing-flow.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    MME->>MME: save the hash of the pairing key<br> and enable desktop app
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:24]:
```suggestion
    Note over MME: - Save the pairing key hash<br/>- Enable desktop app
```

File: docs/totp-pairing-flow.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    User->>MME: "Enable Desktop App" button clicked
+        MME->>MMD: start a limited connection
+        activate MMD
+        Note over MME,MMD: - Create WebSocket<br/>- Encrypt Stream<br/>- Perform Handshake
+        MMD->>MME: return connection established
+        deactivate MMD
+        Note over MME: Generates TOTP
+    MME->>User: display Pairing Page to the user
+    User->>MMD: submit TOTP (6 digits) displayed in the MME
+    MMD->>MME: send TOTP to MME
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:22]: For the arrows, do we really need to include the destination like `MME` if the diagram is illustrating that for us?

File: docs/totp-pairing-flow.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    Note over MME: Restart to run the background<br> process in the MMD
+    
+
+```
+
+## Overview
+
+All starts on the MetaMask extension side whenever a user clicks `Enable Desktop App` in Settings>Experimental.
+
+The MetaMask extension connects via [WebSockets](./encryption.md#communication) with the desktop app and once the streams are created and the handshake performed a 6 digits TOTP using `SHA1` refreshed every 30 seconds and provided by [OTPAuth](https://github.com/hectorm/otpauth) is displayed.
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:37]:
```suggestion
The MetaMask extension connects with the desktop app via [WebSockets](./encryption.md#communication). Once the streams are created and the handshake is completed, a 6 digit TOTP is displayed.

The `SHA1` based TOTP is refreshed every 30 seconds and is generated by [OTPAuth](https://github.com/hectorm/otpauth).
```

File: docs/totp-pairing-flow.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+Every 5 TOTP attempts the secret is renewed.
+
+The user submits the 6 digits and once [TOTP is validated](../packages/common/src/pairing.ts) by the MetaMask Extension a pairing key is generated using AES-GCM with a 256-bit, send to the desktop app alongside a boolean to enable desktop and saved the pairing key hash into the Metamask extension.
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:41]:
```suggestion
After the user enters the 6 digits and the TOTP is [validated](../packages/common/src/pairing.ts) by the MetaMask extension:

- A pairing key is generated and sent to MetaMask Desktop.
- Desktop mode is enabled in the extension.
- The hash of the pairing key is saved in the extension state.
```

File: docs/totp-pairing-flow.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+All starts on the MetaMask extension side whenever a user clicks `Enable Desktop App` in Settings>Experimental.
+
+The MetaMask extension connects via [WebSockets](./encryption.md#communication) with the desktop app and once the streams are created and the handshake performed a 6 digits TOTP using `SHA1` refreshed every 30 seconds and provided by [OTPAuth](https://github.com/hectorm/otpauth) is displayed.
+
+Every 5 TOTP attempts the secret is renewed.
+
+The user submits the 6 digits and once [TOTP is validated](../packages/common/src/pairing.ts) by the MetaMask Extension a pairing key is generated using AES-GCM with a 256-bit, send to the desktop app alongside a boolean to enable desktop and saved the pairing key hash into the Metamask extension. 
+
+After the pairing is concluded whenever the connection starts the extension waits for the desktop pairing key, hashes it and checks if it matches with the hash of the pairing key stored in the extension before establishing the connection and syncing the state.
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:43]:
```suggestion
Once the pairing is completed, whenever a new connection occurs, the extension waits for the desktop pairing key, hashes it, and checks if it matches the stored hash before establishing the authorised connection and syncing the state.
```

File: docs/totp-pairing-flow.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+Once the connection is active, the state is constantly synced between the desktop app and the extension.
```
Comment (matthewwalsh0) [inline docs/totp-pairing-flow.md:45]:
```suggestion
Once the authorised connection is active, the state is continuously synchronised between MetaMask Desktop and the MetaMask extension whenever the persisted state is updated.
```

File: docs/totp-pairing-flow.md (hunk 3 above)
Comment (vinistevam) [reply docs/totp-pairing-flow.md]: Good point, I don't think so. I changed it for simplicity.

**Review by matthewwalsh0 — APPROVED** [review]

**Review by cryptotavares — APPROVED** [review]

**Pull request merged**